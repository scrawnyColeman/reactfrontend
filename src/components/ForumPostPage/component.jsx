import React from 'react';
import styled from 'styled-components';
import { useLocation } from 'react-router-dom';
import ForumPost from '../ForumPost/component';
import forumposts from '../../data/forumposts';
import ForumPostNewCommentForm from '../ForumPostNewCommentForm/component';

const StyledItemWrapper = styled.div`
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-gap: 8px;
    margin: 20vh 20px 0 20px;
`;

const ForumPostPage = () => {
    const location = useLocation();
    const postId = location.pathname.split('/forum/')[1];

    return (
        <StyledItemWrapper>
            <ForumPost id={postId} data={forumposts} />
            <ForumPostNewCommentForm />
        </StyledItemWrapper>
    );
};
export default ForumPostPage;
