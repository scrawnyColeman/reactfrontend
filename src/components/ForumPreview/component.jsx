import { useHistory } from 'react-router-dom';
import React from 'react';
import styled from 'styled-components';
import Button from '../Button/component';
import { colours } from '../../constants/styles';

const StyledContainer = styled.div`
    margin: 0 auto;
    width: 45vw;
    height: 600px;
`;
const StyledPostsContainer = styled.div`
    overflow-y: scroll;
    height: 580px;
`;
const StyledTitle = styled.div`
    padding: 6px;
    color: ${colours.secondary}
    text-align: left;
`;

const StyledWrapper = styled.div`
    display: grid;
    grid-template-rows: 1fr min-content;
    margin: 0 8px;
    text-align: left;
`;
const StyledHead = styled.div`
    display: flex;
    justify-content: space-between;
    color: ${colours.primary};
    font-size: 1.25rem;
`;
const StyledBody = styled.div`
    display flex;
    justify-content: space-between;
    color: ${colours.secondary}
`;

const ForumPreview = ({ onClick, posts }) => {
    const history = useHistory();
    return (
        <StyledContainer>
            <StyledTitle>Forum Posts:</StyledTitle>
            <StyledPostsContainer>
                {posts.map((post) => {
                    const { id, title, type, comments } = post;
                    return (
                        <StyledWrapper key={id}>
                            <StyledHead>
                                <p>{title}</p>
                                <p>[{type.toUpperCase()}]</p>
                            </StyledHead>
                            <StyledBody>
                                <Button
                                    onClick={() => {
                                        history.push({ pathname: `forum/${id}` });
                                        onClick(id);
                                    }}
                                    text="View"
                                    size="small"
                                    variant="outlined"
                                    hierarchy="primary"
                                />
                                <p>{comments.length} comments</p>
                            </StyledBody>
                        </StyledWrapper>
                    );
                })}
            </StyledPostsContainer>
        </StyledContainer>
    );
};

export default ForumPreview;
