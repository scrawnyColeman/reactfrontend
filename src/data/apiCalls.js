import axios from 'axios';

const BASE_URL = 'http://localhost:8080/njoy';
const authToken = sessionStorage.getItem('authToken');
const headers = { headers: { authorization: authToken } };

/****** Course fetchrequests *******/
export const fetchAllLessons = () => axios.get(`${BASE_URL}/lessons`, headers);

export const fetchLessonById = (id) => axios.get(`${BASE_URL}/lessons/${id}`, headers);

export const fetchApprovedLessons = () => axios.get(`${BASE_URL}/lessons/approved`, headers);

export const fetchUnapprovedLessons = () => axios.get(`${BASE_URL}/lessons/unapproved`, headers);

export const fetchUserLessons = (id) => {
    return axios.get(`${BASE_URL}/user_lessons/${id}`, headers);
};

/****** Forum fetch requests *******/
export const fetchAllForumPosts = () => axios.get(`${BASE_URL}/forum_posts`, headers);

export const fetchForumPostByLesson = (id) => axios.get(`${BASE_URL}/forum_posts/lesson/${id}`, headers);

export const fetchForumPost = (id) => axios.get(`${BASE_URL}/forum_posts/${id}`, headers);

export const fetchForumPostComments = (id) => axios.get(`${BASE_URL}/comments/post/${id}`, headers);

/****** Practicals fetch requests *******/
export const fetchPracticalByLessonId = (id) => axios.get(`${BASE_URL}/practicals/lesson/${id}`, headers);

export const fetchSolutionByPracticalId = (id) => axios.get(`${BASE_URL}/solutions/practical/${id}`, headers);

/****** Quizzes fetch requests *******/
export const fetchAllQuestionsByLessonId = (id) => axios.get(`${BASE_URL}/questions/lesson/${id}`, headers);

export const fetchAllAnswersByLessonId = (id) => axios.get(`${BASE_URL}/questions/answers/lesson/${id}`, headers);

/****** Languages fetch requests *******/
export const fetchAllLanguages = () => axios.get(`${BASE_URL}/languages`, headers);

/****** Users fetch requests *******/
export const fetchUserByUsernameOrEmail = (username, email) =>
    axios.get(`${BASE_URL}/user/${username}/${email}`, headers);

/****** Users update requests *******/
export const updateUserInfo = (userId, username, email) =>
    axios.put(`${BASE_URL}/users/${userId}/username/${username}/email/${email}`, headers);

/****** Users post requests *******/
export const setUserLesson = (userId, lessonId) => axios.post(`${BASE_URL}/user/${userId}/lesson/${lessonId}`, headers);

export const createForumPost = (post) => axios.post(`${BASE_URL}/forum_posts`, post, headers);

export const createComment = (comment, userId, postId) =>
    axios.post(`${BASE_URL}/comments/${userId}/${postId}`, comment, headers);

export const deleteComment = (commentId) => axios.put(`${BASE_URL}/comments/delete/${commentId}`, headers);

export const editComment = (commentId, comment) =>
    axios.put(`${BASE_URL}/comments/edit/${commentId}`, comment, headers);

export const deleteLesson = (lessonId) => axios.delete(`${BASE_URL}/lessons/delete/${lessonId}`, headers);

export const approveLesson = (lessonId) => axios.put(`${BASE_URL}/lessons/approve/${lessonId}`, headers);

export const postLesson = (lesson, languageId) => axios.post(`${BASE_URL}/lessons/new/${languageId}`, lesson, headers);

export const postQuestions = (questions, lessonId) =>
    axios.post(`${BASE_URL}/lessons/questions/new/${lessonId}`, questions, headers);

export const postAnswers = (answers, questionId) =>
    axios.post(`${BASE_URL}/lessons/answers/new/${questionId}`, answers, headers);

export const postPractical = (practical, lessonId) =>
    axios.post(`${BASE_URL}/lessons/practical/new/${lessonId}`, practical, headers);

export const postSolution = (solution, practicalId) =>
    axios.post(`${BASE_URL}/lessons/solutions/new/${practicalId}`, solution, headers);

export const setLessonComplete = (lessonId, userId) =>
    axios.put(`${BASE_URL}/lesson/complete/${lessonId}/${userId}`, headers);
