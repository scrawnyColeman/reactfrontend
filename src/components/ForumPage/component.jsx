import React, { useState } from 'react';
import { useLocation } from 'react-router-dom';
import styled from 'styled-components';
import ForumPreview from '../ForumPreview/component';
import ForumPostPage from '../ForumPostPage/component';
import data from '../../data/forumposts.js';
import CreateForumPost from '../CreateForumPost/component';

const StyledItemWrapper = styled.div`
    display: grid;
    grid-template-columns: 1fr 1fr;
    margin: 20vh 20px 0 20px;
`;

const ForumPage = () => {
    const location = useLocation().search;
    const [param, setParam] = useState(location);
    return (
        <StyledItemWrapper>
            {Boolean(param) ? (
                <ForumPostPage />
            ) : (
                <>
                    <ForumPreview posts={data.posts} onClick={setParam} />
                    <CreateForumPost />
                </>
            )}
        </StyledItemWrapper>
    );
};
export default ForumPage;
