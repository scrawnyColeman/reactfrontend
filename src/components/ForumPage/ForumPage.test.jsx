import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-styled-components';
import React from 'react';
import ForumPage from './component';
import { useLocation } from 'react-router-dom';
import { fetchAllForumPosts } from '../../data/apiCalls.js';
import ForumPreview from '../ForumPreview/component';
import ForumPostPage from '../ForumPostPage/component';

jest.mock('../../data/apiCalls.js', () => ({
    fetchAllForumPosts: jest.fn(),
}));

jest.mock('react-router-dom', () => {
    const useLocation = jest.fn(() => ({ pathname: 'something', search: 'somewhere' }));
    return { useLocation };
});

Enzyme.configure({ adapter: new Adapter() });

describe('ForumPage component', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    test('component matches snapshot with URL param', () => {
        fetchAllForumPosts.mockResolvedValueOnce({ data: { post: { id: 1 } } });
        const wrapper = shallow(<ForumPage />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(ForumPreview).length).toBe(0);
        expect(wrapper.find(ForumPostPage).length).toBe(1);
    });
    test('component matches snapshot with no URL param', () => {
        useLocation.mockReturnValue({ search: null });
        fetchAllForumPosts.mockResolvedValueOnce({ data: { post: { id: 1 } } });
        const wrapper = shallow(<ForumPage />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(ForumPreview).length).toBe(1);
        expect(wrapper.find(ForumPostPage).length).toBe(0);
    });
});
