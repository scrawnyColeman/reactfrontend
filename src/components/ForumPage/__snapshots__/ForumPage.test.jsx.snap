// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForumPage component component matches snapshot with URL param 1`] = `
<styled.div>
  <ForumPreview
    onClick={[Function]}
    posts={
      Array [
        Object {
          "comments": Array [
            Object {
              "authorusername": "anon",
              "comment": "No it can't my guy",
              "commentid": "1",
              "parentid": "0",
            },
            Object {
              "comment": "Oh okay. That sucks",
              "commentid": "2",
              "parentid": "1",
            },
            Object {
              "comment": "Helpful post. Good Job... OP, you can do it by...",
              "commentid": "3",
              "parentid": "1",
            },
            Object {
              "comment": "Thank you!!",
              "commentid": "4",
              "parentid": "3",
            },
          ],
          "id": "1",
          "question": "So I understand that for different situations, either the ArrayList API or the LinkedList API will be more efficient. So what should I look for in terms of design that will tell me which one of those structures is recommended? Is it more related to preference rather than specific uses?",
          "title": "ArrayLists vs Linked Lists",
          "type": "HELP",
        },
        Object {
          "comments": Array [
            Object {
              "comment": "No it can't my guy",
              "commentid": "5",
              "parentid": "0",
            },
            Object {
              "comment": "Oh okay. That sucks",
              "commentid": "6",
              "parentid": "5",
            },
            Object {
              "comment": "Helpful post. Good Job smh... OP, you can do it by...",
              "commentid": "7",
              "parentid": "5",
            },
            Object {
              "comment": "Thank you!!",
              "commentid": "8",
              "parentid": "7",
            },
          ],
          "id": "2",
          "question": "Hey all, I'm currently working my way through the MOOC course and I've just completed an exercise where a scanner asks for a list of numbers from the user, and then it prints out what the smallest number in the index is, and prints out every index it is at. I'm wondering if this could be completed in just one loop.",
          "title": "Can this be done in one loop?",
          "type": "HELP",
        },
        Object {
          "comments": Array [],
          "id": "3",
          "question": "This thread is for discussing the ReactJS lesson, taught in JavaScript",
          "title": "ReactJS [JavaScript]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "4",
          "question": "This thread is for discussing the Lambda Functions lesson, taught in Java",
          "title": "Lambda Functions [Java]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "5",
          "question": "This thread is for discussing the Recursion lesson, taught in Python",
          "title": "Recursion [Python]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "6",
          "question": "This thread is for discussing the Polymorphism lesson, taught in Java",
          "title": "Polymorphism [Java]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "7",
          "question": "This thread is for discussing the Control Flow and Conditionals lesson, taught in Python",
          "title": "Control Flow and Conditionals [Python]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "8",
          "question": "This thread is for discussing the Classes and Objects lesson, taught in Python",
          "title": "Classes and Objects [Python]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "9",
          "question": "This thread is for discussing the ES6 Arrow Functions lesson, taught in JavaScript",
          "title": "ES6 Arrow Functions [JavaScript]",
          "type": "LESSON",
        },
      ]
    }
  />
  <CreateForumPost />
</styled.div>
`;

exports[`ForumPage component component matches snapshot with URL param 2`] = `
<styled.div>
  <ForumPreview
    onClick={[Function]}
    posts={
      Array [
        Object {
          "comments": Array [
            Object {
              "authorusername": "anon",
              "comment": "No it can't my guy",
              "commentid": "1",
              "parentid": "0",
            },
            Object {
              "comment": "Oh okay. That sucks",
              "commentid": "2",
              "parentid": "1",
            },
            Object {
              "comment": "Helpful post. Good Job... OP, you can do it by...",
              "commentid": "3",
              "parentid": "1",
            },
            Object {
              "comment": "Thank you!!",
              "commentid": "4",
              "parentid": "3",
            },
          ],
          "id": "1",
          "question": "So I understand that for different situations, either the ArrayList API or the LinkedList API will be more efficient. So what should I look for in terms of design that will tell me which one of those structures is recommended? Is it more related to preference rather than specific uses?",
          "title": "ArrayLists vs Linked Lists",
          "type": "HELP",
        },
        Object {
          "comments": Array [
            Object {
              "comment": "No it can't my guy",
              "commentid": "5",
              "parentid": "0",
            },
            Object {
              "comment": "Oh okay. That sucks",
              "commentid": "6",
              "parentid": "5",
            },
            Object {
              "comment": "Helpful post. Good Job smh... OP, you can do it by...",
              "commentid": "7",
              "parentid": "5",
            },
            Object {
              "comment": "Thank you!!",
              "commentid": "8",
              "parentid": "7",
            },
          ],
          "id": "2",
          "question": "Hey all, I'm currently working my way through the MOOC course and I've just completed an exercise where a scanner asks for a list of numbers from the user, and then it prints out what the smallest number in the index is, and prints out every index it is at. I'm wondering if this could be completed in just one loop.",
          "title": "Can this be done in one loop?",
          "type": "HELP",
        },
        Object {
          "comments": Array [],
          "id": "3",
          "question": "This thread is for discussing the ReactJS lesson, taught in JavaScript",
          "title": "ReactJS [JavaScript]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "4",
          "question": "This thread is for discussing the Lambda Functions lesson, taught in Java",
          "title": "Lambda Functions [Java]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "5",
          "question": "This thread is for discussing the Recursion lesson, taught in Python",
          "title": "Recursion [Python]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "6",
          "question": "This thread is for discussing the Polymorphism lesson, taught in Java",
          "title": "Polymorphism [Java]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "7",
          "question": "This thread is for discussing the Control Flow and Conditionals lesson, taught in Python",
          "title": "Control Flow and Conditionals [Python]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "8",
          "question": "This thread is for discussing the Classes and Objects lesson, taught in Python",
          "title": "Classes and Objects [Python]",
          "type": "LESSON",
        },
        Object {
          "comments": Array [],
          "id": "9",
          "question": "This thread is for discussing the ES6 Arrow Functions lesson, taught in JavaScript",
          "title": "ES6 Arrow Functions [JavaScript]",
          "type": "LESSON",
        },
      ]
    }
  />
  <CreateForumPost />
</styled.div>
`;
